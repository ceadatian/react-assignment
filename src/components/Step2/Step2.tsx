/**
 * requirement
 * a dropdown menu for selecting the restaurant based on the meal category selected in step 1.
 * pass the selected restaurant data to the next step.
 */

import React, { useState } from 'react';
import { Form, Select, Button } from 'antd';
import './Step2.css'

interface Props {
  availableRestaurants: Restaurant[];
  onSubmit: (selectedRestaurant: Restaurant) => void;
	onBack: () => void;
}

interface Restaurant {
  id: number;
  name: string;
  restaurant: string;
  availableMeals: string[];
}

const Step2: React.FC<Props> = ({ availableRestaurants, onSubmit, onBack }) => {
  const [selectedRestaurant, setSelectedRestaurant] = useState<Restaurant | undefined>();

	// const handleRestaurantChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
  //   const selectedRestaurantId = event.target.value;
  //   const selectedRestaurant = availableRestaurants.find(
  //     (restaurant) => restaurant.id === +selectedRestaurantId
  //   );
  //   setSelectedRestaurant(selectedRestaurant);
  // };
  const handleRestaurantChange = (value: string) => {
    const selectedRestaurantId = value;
    const selectedRestaurant = availableRestaurants.find(
      (restaurant) => restaurant.id === +selectedRestaurantId
    );
    setSelectedRestaurant(selectedRestaurant);
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    // event.preventDefault();
    if (selectedRestaurant) {
      onSubmit(selectedRestaurant);
    }
  };

  const uniqueRestaurants = availableRestaurants.filter((item, index, self) => self.map(i => i.restaurant).indexOf(item.restaurant) === index);

  return (
    <Form className='global-step' onFinish={handleSubmit}>
      <Form.Item label="Please Select a Restaurant:" name="selectedRestaurantId" style={{ width: '400px'}}>
        <Select onChange={handleRestaurantChange}>
          <Select.Option value="">Please Select a Restaurant</Select.Option>
          {uniqueRestaurants.map((restaurant) => (
            <Select.Option key={restaurant.id} value={restaurant.id}>
              {restaurant.restaurant}
            </Select.Option>
          ))}
        </Select>
      </Form.Item>
			<Form.Item>
        <Button className='global-previous' onClick={onBack}>Previous</Button>
        <Button type="primary" htmlType="submit" disabled={!selectedRestaurant}>Next</Button>
      </Form.Item>
    </Form>
  );
};

export default Step2;
