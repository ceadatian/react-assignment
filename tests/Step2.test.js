import React from "react";
import { shallow } from "enzyme";
import Step2 from "../src/components/Step2/Step2";

describe("Step2 component", () => {
  it("renders correctly", () => {
    const wrapper = shallow(
      <Step2 availableRestaurants={[]} onSubmit={() => {}} onBack={() => {}} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onSubmit function with correct argument when form is submitted", () => {
    const onSubmit = jest.fn();
    const availableRestaurants = [
      {
        id: 1,
        name: "Dish 1",
        restaurant: "Restaurant 1",
        availableMeals: ["breakfast"],
      },
      {
        id: 2,
        name: "Dish 2",
        restaurant: "Restaurant 2",
        availableMeals: ["lunch"],
      },
      {
        id: 3,
        name: "Dish 3",
        restaurant: "Restaurant 3",
        availableMeals: ["dinner"],
      },
    ];
    const wrapper = shallow(
      <Step2
        availableRestaurants={availableRestaurants}
        onSubmit={onSubmit}
        onBack={() => {}}
      />
    );
    const restaurantSelect = wrapper.find("select");
    restaurantSelect.simulate("change", { target: { value: "2" } });
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });
    expect(onSubmit).toHaveBeenCalledWith({
      id: 2,
      name: "Dish 2",
      restaurant: "Restaurant 2",
      availableMeals: ["lunch"],
    });
  });

  it("should call onBack function when the back button is clicked", () => {
    const onBack = jest.fn();
    const wrapper = shallow(
      <Step2 availableRestaurants={[]} onSubmit={() => {}} onBack={onBack} />
    );
    wrapper.find("button.global-previous").simulate("click");
    expect(onBack).toHaveBeenCalled();
  });

  it("should disable the submit button when no restaurant is selected", () => {
    const wrapper = shallow(
      <Step2 availableRestaurants={[]} onSubmit={() => {}} onBack={() => {}} />
    );
    expect(wrapper.find('button[type="submit"]').prop("disabled")).toBe(true);
    wrapper.setProps({
      availableRestaurants: [
        {
          id: 1,
          name: "Dish 1",
          restaurant: "Restaurant 1",
          availableMeals: ["breakfast"],
        },
      ],
    });
    expect(wrapper.find('button[type="submit"]').prop("disabled")).toBe(true);
    wrapper.find("select").simulate("change", { target: { value: "1" } });
    expect(wrapper.find('button[type="submit"]').prop("disabled")).toBe(false);
  });
});
