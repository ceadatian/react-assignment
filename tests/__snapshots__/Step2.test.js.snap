// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step2 component renders correctly 1`] = `ShallowWrapper {}`;
